torcheeg/models/transformer/retnet.py

```diff
@@ -0,0 +1,359 @@
+        self.head_size = hidden_size // heads
+        #self.head_v_dim = hidden_size * 2 if double_v_dim else hidden_size
+        
+        self.gammas = (1 - torch.exp(torch.linspace(math.log(1/32), math.log(1/512), heads))).detach().cpu().tolist()
+
+        self.swish = lambda x: x * torch.sigmoid(x)
+        self.W_G = nn.Parameter(torch.randn(hidden_size, self.v_dim) / hidden_size)
+        self.W_O = nn.Parameter(torch.randn(self.v_dim, hidden_size) / hidden_size)
+        self.group_norm = nn.GroupNorm(heads, self.v_dim)
+
+        self.retentions = nn.ModuleList([
+            SimpleRetention(self.hidden_size, gamma, self.head_size, double_v_dim) for gamma in self.gammas
+        ])
+
+    def forward(self, X):
+        """
+        parallel representation of the multi-scale retention mechanism
+        """
+
+        # apply each individual retention mechanism to X
+        Y = []
+        for i in range(self.heads):
+            Y.append(self.retentions[i](X))
+        
+        Y = torch.cat(Y, dim=2)
+        Y_shape = Y.shape
+        Y = self.group_norm(Y.reshape(-1, self.v_dim)).reshape(Y_shape)
+
+        return (self.swish(X @ self.W_G) * Y) @ self.W_O
+    
+    def forward_recurrent(self, x_n, s_n_1s, n):
+        """
+        recurrent representation of the multi-scale retention mechanism
+        x_n: (batch_size, 1, hidden_size)
+        s_n_1s: (batch_size, heads, head_size, head_size)
+
+        """
+    
+        # apply each individual retention mechanism to a slice of X
+        Y = []
+        s_ns = []
+        for i in range(self.heads):
+            y, s_n = self.retentions[i].forward_recurrent(
+                x_n[:, :, :], s_n_1s[i], n
+                )
+            Y.append(y)
+            s_ns.append(s_n)
+        
+        Y = torch.cat(Y, dim=2)
+        Y_shape = Y.shape
+        Y = self.group_norm(Y.reshape(-1, self.v_dim)).reshape(Y_shape)
+        
+        return (self.swish(x_n @ self.W_G) * Y) @ self.W_O, s_ns
+
+    def forward_chunkwise(self, x_i, r_i_1s, i):
+        """
+        chunkwise representation of the multi-scale retention mechanism
+        x_i: (batch_size, chunk_size, hidden_size)
+        r_i_1s: (batch_size, heads, head_size, head_size)
+        """
+        batch, chunk_size, _ = x_i.shape
+
+        # apply each individual retention mechanism to a slice of X
+        Y = []
+        r_is = []
+        for j in range(self.heads):
+            y, r_i = self.retentions[j].forward_chunkwise(
+                x_i[:, :, :], r_i_1s[j], i
+                )
+            Y.append(y)
+            r_is.append(r_i)
+        
+        
+        Y = torch.cat(Y, dim=2)
+        Y_shape = Y.shape
+        Y = self.group_norm(Y.reshape(-1, self.v_dim)).reshape(Y_shape)
+
+        return (self.swish(x_i @ self.W_G) * Y) @ self.W_O, r_is
+
+
+class RetNet(nn.Module):
+    def __init__(self, layers, hidden_dim, ffn_size, heads, double_v_dim=False):
```

Comment:
Please add detailed documentation. For your reference, please see: https://github.com/torcheeg/torcheeg/blob/main/torcheeg/models/transformer/conformer.py#L160-L204

It should contain the Paper, URL, Related Project (to follow further updates), a recommended suite, and an introduction to the args.